### PR TITLE
HTTP PUT output support

### PR DESCRIPTION
**Pull request by termoose**
Experimental feature support for HTTP ingest.

- Tested with ingesting encrypted DASH segments against Akamai MSL4 EP's.
- Tested with flaky ingests, disconnecting and reconnecting.

Relates and hopefully closes issue #149.

### CODE REVIEW

**Comment by amotl** [comment]
Dear Ole Andre,

thanks again for taking care about this. Will you also add 8ddbbf0 into the mix?

With kind regards,
Andreas.

**Comment by amotl** [comment]
@googlebot I signed it!

**Comment by termoose** [comment]
@googlebot I signed it!

**Comment by kqyang** [comment]
@termoose @amotl Thanks for putting this PR together. I did a manual search in CLA database, but I was not able to find CLA associated with your Git committer emails too. Note that you need to use the same email as in your CLA. See the instructions here: https://github.com/google/shaka-packager/pull/737#issuecomment-603203886.

**Comment by termoose** [comment]
@kqyang oh, I got a message from legal saying it was signed. Will double-check it sorry about that

**Comment by amotl** [comment]
@googlebot I signed it!

**Comment by amotl** [comment]
@kqyang I've added andreas.motl@elmyra.de to the list of my Google alternate email addresses and updated the CLA.

**Comment by kqyang** [comment]
@amotl Yes, your CLA is good now.

@termoose Do you know what email address your legal is using for the CLA? Make sure your git email address is included in the CLA.

**Comment by termoose** [comment]
Still waiting for feedback on the CLA here unfortunately, will update when more information is available!

**Comment by kqyang** [comment]
@termoose That's ok. Thanks for the effort. Do you want us to code review while waiting for your legal to sort out the CLA?

**Comment by termoose** [comment]
For sure, totally forgot that this was in draft! Legal just confirmed that CLA should have been signed, hopefully it appears in the CLA database

**Comment by termoose** [comment]
Hi @kqyang, sorry for the standstill on this. Are you able to share which `*@sportradar.com` email has been CLA approved, if any? What I keep getting back from our legal is that it has already been signed, but I don't know under which email. The CLA has been through the hoops here in the company, and I believe it might have been signed along the way.

**Comment by kqyang** [comment]
@termoose Unfortunately the system does not allow me to search by a partial email address. Let me ask internally to see if there is anyway to find that out.

**Comment by kqyang** [comment]
@termoose Here is the reply I got:

> There are no CLAs for SportRadar in our system. 
(And nothing matching /radar/ at all, and /sport/ never as it's own word.)

**Comment by termoose** [comment]
Thanks for checking! I have resubmitted the CLA (looks like it timed out, because I wasn't able to before), this time it should reach the correct person. Hopefully it's resolved when they are back from vacation next week!

**Comment by amotl** [comment]
Hi there,

now I am the person who would like to humbly ask whether merging this feature has made any progress on the legal aspects re. CLA? Apart from that, it looks like the branch has diverged so much that there are some conflicts. Will you be able to rebase this onto master and verify everything still works, @termoose?

Thanks already and with kind regards,
Andreas.

**Comment by termoose** [comment]
I submitted a CLA just for myself since my employer is for some reason not able to sign properly. I will look into getting this rebased and ready for merge this week or the next as we are getting close to using this in production as well now.

**Comment by termoose** [comment]
Merged `master` into my branch, seems like most of the conflicts were that you're now using a `BufferWriter`. I ran a test with MSL4 Akamai push of HLS and DASH, both with DRM (BuyDRM), and it plays back just fine here. Let me know if something is missing or wrongly merged! I don't know what to do about the @googlebot I consent. part above. Is this an issue related to my CLA signing?

**Comment by kqyang** [comment]
> I don't know what to do about the @googlebot I consent. part above. Is this an issue related to my CLA signing?

Emm... I am not sure if it is because of yours CLA signing, but the bot was complaining "one or more commits were authored or co-authored by someone other than the pull request submitter."

I'll see if I can find out.

Thanks for your effort on this PR!

**Comment by amotl** [comment]
Hi again,

> seems like most of the conflicts were that you're now using a BufferWriter

AppVeyor shows different kinds of errors [1]. Is this related somehow?

> Is [there still] an issue related to my CLA signing?

@kqyang promised to check. Maybe this is related to these two commits by Rintaro Kuroiwa (@rkuroiwa) somehow? At least, both commits are not directly associated with his pen.

image

With kind regards,
Andreas.

[1] https://ci.appveyor.com/project/shaka/shaka-packager/builds/35724371/job/d3va1bvda78ago4o

**Comment by kqyang** [comment]
> @kqyang promised to check. Maybe this is related to these two commits by Rintaro Kuroiwa (@rkuroiwa) somehow? At least, both commits are not directly associated with his pen.

We are still checking. Apologize for the delay. 

> AppVeyor shows different kinds of errors [1]. Is this related somehow?

It looks like a problem in the test.

```c++
TEST_F(HttpFileTest, PutChunkedTranser) {
  std::unique_ptr<File, FileCloser> writer(
      File::Open("put+http://127.0.0.1:8080/test_out", "w"));
  ASSERT_TRUE(writer);
  ASSERT_EQ(kWriteBufferSize, writer->Write(kWriteBuffer, kWriteBufferSize));
  writer.release()->Close();
}
```

```shell
[ RUN      ] HttpFileTest.PutChunkedTranser
[1013/131933:ERROR:local_file.cc(85)] Failed to create directory put+http: ErrorCode 123
c:\projects\shaka-packager\src\packager\file\http_file_unittest.cc(20): error: Value of: writer
  Actual: false
Expected: true
[  FAILED  ] HttpFileTest.PutChunkedTranser (27 ms)
```

Should the URL of the file open just be "http://127.0.0.1:8080/test_out"?

Also, a separate question, do we support HTTP methods other than "put"?

**Comment by kqyang** [comment]
@amotl Here is the problem:

> amotl | andreas.motl@elmyra.de | need_author_consent

We need your consent for this PR. See the instruction at https://github.com/google/shaka-packager/pull/737#issuecomment-701226448, i.e.

> Please have them confirm that by leaving a comment that contains only @googlebot I consent

**Comment by amotl** [comment]
Oh, it was me! I believed I had solved this properly already. Thanks!

**Comment by amotl** [comment]
@googlebot I consent

**Comment by termoose** [comment]
> It looks like a problem in the test.
> 
> ```c++
> TEST_F(HttpFileTest, PutChunkedTranser) {
>   std::unique_ptr<File, FileCloser> writer(
>       File::Open("put+http://127.0.0.1:8080/test_out", "w"));
>   ASSERT_TRUE(writer);
>   ASSERT_EQ(kWriteBufferSize, writer->Write(kWriteBuffer, kWriteBufferSize));
>   writer.release()->Close();
> }
> ```
> 
> ```shell
> [ RUN      ] HttpFileTest.PutChunkedTranser
> [1013/131933:ERROR:local_file.cc(85)] Failed to create directory put+http: ErrorCode 123
> c:\projects\shaka-packager\src\packager\file\http_file_unittest.cc(20): error: Value of: writer
>   Actual: false
> Expected: true
> [  FAILED  ] HttpFileTest.PutChunkedTranser (27 ms)
> ```
> 
> Should the URL of the file open just be "http://127.0.0.1:8080/test_out"?

That's a strange test, but I'm guessing it should be just http since that's the only prefix we're checking for with the `kHttpFilePrefix` and `kHttpsFilePrefix`.
It does run for me though, which doesn't make sense at all?

```
[----------] 1 test from HttpFileTest
[ RUN      ] HttpFileTest.PutChunkedTranser
[       OK ] HttpFileTest.PutChunkedTranser (20 ms)
[----------] 1 test from HttpFileTest (20 ms total)
```

It gives me a 404 if I change the test to `http`
```
[----------] 1 test from HttpFileTest
[ RUN      ] HttpFileTest.PutChunkedTranser
[1027/094205:ERROR:http_file.cc(218)] PUT request for http://127.0.0.1:8080/test_out failed. Reason: HTTP response code said error. Response code: 404.
[       OK ] HttpFileTest.PutChunkedTranser (47 ms)
[----------] 1 test from HttpFileTest (47 ms total)
```
It makes sense that the test is not failing anyway since it's only asserting `writer` and checking that `kWriteBufferSize` is equal to bytes written, but `Writer` writes to a `cache_` buffer before doing the http request, so that will probably never fail. I made the change to `http` anyway.

> Also, a separate question, do we support HTTP methods other than "put"?

It's only doing PUT for now but I see there's code for other methods added in the large commit above, does not seem to be in use.

**Comment by amotl** [comment]
Hi there,

> It gives me a 404 if I change the test to `http`.

It will not be so easy. However, I can't remember the intrinsic details that took me to that decision to add the prefix like that besides only supporting PUT. I believe it has something to do with that I had to use something like the `put+` prefix to detect that the output should go to a non-file handler.

Will you be able to make it work after sublimating yourself into the code again, @termoose?

With kind regards,
Andreas.

**Comment by amotl** [comment]
The innocent and rather superfluous code like 4491dee43 ff. has been there for a reason to be able to tell different kinds of files apart.

Otherwise, the machinery doesn't know to handle that as HTTP and will try to access the local file system
```
[1013/131933:ERROR:local_file.cc(85)] Failed to create directory put+http: ErrorCode 123
```

**Comment by amotl** [comment]
Is it actually a good sign that _AppVeyor build succeeded_ right now? So, will that be ready to merge?

**Comment by termoose** [comment]
Ah I didn't see your `GetEffectiveFileAddress` function! Could it be that there are other mechanism at work in the "loosely coupled subsystem" now that makes it handle the `http` protocol? Whenever I tested with ingesting to Akamai MSL4 I didn't specify the output gateway to have `put+http://` BTW. I will test the branch once again against Akamai's gateways for DASH and HLS and see how it behaves. This merge is getting complicated :)

But yes the test passes, so that's a bit strange, but I don't see how it could fail actually. The only way I see this test failing is if the `writer` is `nullptr`, and for some strange reason it was when the test was running with `put+http`. I guessing that this is because `put+http` is not in any of the elements in `kFileTypeInfo[]` and the construction fails? Anyway enough rabling from me! I'll run some tests tomorrow and see how it behaves. I'd also like to add support for multiple output gateways at some point so we can push to a backup, but that's for another time!

**Comment by amotl** [comment]
Dear Ole Andre,

> Could it be that there are other mechanism at work in the "loosely coupled subsystem" now that makes it handle the http protocol?

That might well be the case. Magic happens sometimes. ;] Maybe also I was just too stupid back then. Thanks for the fix through a1e5a40. If we need to support other HTTP methods than PUT, we will be able to revisit this later.

> I will test the branch once again against Akamai's gateways for DASH and HLS and see how it behaves.

Thank you so much! I believe if that will succeed, we will be ready to merge and address other things related to the tests later.

With kind regards,
Andreas.

File: packager/file/http_file.cc
Comment (TheModMaker) [inline packager/file/http_file.cc:183]: The URL is in a field, so no need to pass as an argument.

**Review by TheModMaker — CHANGES_REQUESTED** [review]

File: docs/source/tutorials/http_upload.rst
Comment (amotl) [reply docs/source/tutorials/http_upload.rst:123]: That's just a leftover from the time it hasn't been supported. Thanks for also spotting this!

**Comment by amotl** [comment]
Dear Jacob,

> @TheModMaker requested changes [8 hours ago](https://github.com/google/shaka-packager/pull/737#pullrequestreview-524713801).

Thanks a bunch for your thorough review. Sorry that most of these are coming from the fact that I am not a native C++ speaker. I also want to apologize that I am currently busy with other obligations, so I am humbly asking if you will be able to address these adjustments, @termoose?

With kind regards,
Andreas.

**Comment by termoose** [comment]
Thanks for the very thorough review @TheModMaker! I just started at the top and can continue when time allows, think we should be able to get through these. Most of these changes are not done by me so it's a bit challenging to amend them.

**Comment by amotl** [comment]
Hi again,

I just wanted to ping you about this again. Because @termoose stated:
> Most of these changes are not done by me so it's a bit challenging to amend them.

I believe most of them are coming from my pen while working on this the other day. However, as I currently don't see any chance to continue working on those, I want to humbly ask if one of the core developers might be able to take over addressing the issues reported by @TheModMaker?

With kind regards,
Andreas.

**Comment by CaitlinOCallaghan** [comment]
Hi everyone, @amotl @termoose @TheModMaker @joeyparrish! I was wondering what the status is on this PR? My work uses the HTTP upload feature, and I would love to see it get pulled into master soon. Please let me know if I can help in any way. 

Cheers,
Caitlin

**Comment by TheModMaker** [comment]
We've decided to merge this as-is and I'll make the requested changes.  Thanks for the contribution.

**Comment by amotl** [comment]
Hi Jacob,

thank you very much for integrating this feature and for taking over development.

With kind regards,
Andreas.

**Pull request merged**
